src/currentprocess/terminalsource.rs

```diff
@@ -0,0 +1,231 @@
+use std::io::{self, Result, Write};
+use std::sync::{Arc, Mutex, MutexGuard};
+
+pub(crate) use termcolor::Color;
+use termcolor::{ColorChoice, ColorSpec, StandardStream, WriteColor};
+
+use crate::process;
+
+/// Allows you to write to a terminal or other output stream with a lock.
+/// This is a stand-in for std::io::StdoutLock.
+pub trait WriterLock: Write {}
+
+/// This is a stand-in for std::io::Stdout or std::io::Stderr.
+pub trait Writer: Send + Sync {
+    /// Locks the terminal or other output stream.
+    /// This is a stand-in for std::io::Stdout::lock or std::io::Stderr::lock.
+    fn lock(&self) -> Box<dyn WriterLock + '_>;
+}
+
+/// Satisfies `Writer`, and supports optional coloring
+/// to either of the standard output streams, stdout and stderr.
+pub trait Terminal: Writer {
+    fn fg(&mut self, color: Color) -> io::Result<()>;
+    fn bg(&mut self, color: Color) -> io::Result<()>;
+    fn attr(&mut self, attr: Attr) -> io::Result<()>;
+    fn reset(&mut self) -> io::Result<()>;
+    fn carriage_return(&mut self) -> io::Result<()>;
+}
+
+// -------------- stdout -------------------------------
+
+/// Stand-in for std::io::stdout.
+pub trait StdoutTerminalSource {
+    fn stdout(&self) -> Box<dyn Terminal>;
+}
+
+// -------------- stderr -------------------------------
+
+/// Stand-in for std::io::stderr.
+pub trait StderrTerminalSource {
+    fn stderr(&self) -> Box<dyn Terminal>;
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum Attr {
+    Bold,
+    ForegroundColor(Color),
+}
+
+/// Disable all terminal controls on non-tty's.
+/// A wrapper around a `StandardStream` that provides a `Writer` implementation.
+struct StandardStreamTerminal {
+    stream: StandardStream,
+    color: ColorSpec,
+}
+
+impl Terminal for StandardStreamTerminal {
+    fn fg(&mut self, color: Color) -> io::Result<()> {
+        if !self.stream.supports_color() {
+            return Ok(());
+        }
+        self.color.set_fg(Some(color));
+        self.stream.set_color(&self.color)
+    }
+
+    fn bg(&mut self, color: Color) -> io::Result<()> {
+        if !self.stream.supports_color() {
+            return Ok(());
+        }
+        self.color.set_bg(Some(color));
+        self.stream.set_color(&self.color)
+    }
+
+    fn attr(&mut self, attr: Attr) -> io::Result<()> {
+        if !self.stream.supports_color() {
+            return Ok(());
+        }
+        match attr {
+            Attr::Bold => self.color.set_bold(true),
+            Attr::ForegroundColor(color) => self.color.set_fg(Some(color)),
+        };
+        self.stream.set_color(&self.color)
+    }
+
+    fn reset(&mut self) -> io::Result<()> {
+        if !self.stream.supports_color() {
+            return Ok(());
+        }
+        self.stream.reset()
+    }
+
+    fn carriage_return(&mut self) -> io::Result<()> {
```

Comment:
I think we should drop this method. It existed because the other terminal abstraction we used provided it, but if termcolor just wants a \r in the stream, we can just make our code do that directly. 

This can be done in a later PR to keep this one focused on the like-for-like exchange.

Reply:
I will change it in my next PR.